frame/contracts/src/rent.rs

```diff
@@ -127,75 +147,140 @@ fn try_evict_or_and_pay_rent<T: Trait>(
+				deduct_block: current_block_number,
+				..alive_contract_info
+			});
+			<ContractInfoOf<T>>::insert(account, &contract_info);
 
-		return (RentOutcome::Ok, Some(contract_info));
+			amount.withdraw(account);
+			Some(contract_info)
+		}
 	}
-
-	(RentOutcome::Ok, Some(ContractInfo::Alive(contract)))
 }
 
 /// Make account paying the rent for the current block number
 ///
-/// NOTE: This function acts eagerly.
-pub fn pay_rent<T: Trait>(account: &T::AccountId) -> Option<ContractInfo<T>> {
-	try_evict_or_and_pay_rent::<T>(account, Zero::zero(), true).1
+/// NOTE this function performs eviction eagerly. All changes are read and written directly to
+/// storage.
+pub fn collect_rent<T: Trait>(account: &T::AccountId) -> Option<ContractInfo<T>> {
+	let contract_info = <ContractInfoOf<T>>::get(account);
+	let alive_contract_info = match contract_info {
+		None | Some(ContractInfo::Tombstone(_)) => return contract_info,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+
+	let current_block_number = <frame_system::Module<T>>::block_number();
+	let verdict = consider_case::<T>(
+		account,
+		current_block_number,
+		Zero::zero(),
+		&alive_contract_info,
+	);
+	enact_verdict(account, alive_contract_info, current_block_number, verdict)
 }
 
-/// Evict the account if it should be evicted at the given block number.
+/// Process a snitch that a contract under the given address should be evicted.
 ///
-/// `handicap` gives a way to check or pay the rent up to a moment in the past instead
+/// Enact the eviction right away if the contract should be evicted and return true.
+/// Otherwise, **do nothing** and return false.
+///
+/// The `handicap` parameter gives a way to check the rent to a moment in the past instead
 /// of current block. E.g. if the contract is going to be evicted at the current block,
-/// `handicap=1` can defer the eviction for 1 block.
+/// `handicap = 1` can defer the eviction for 1 block. This is useful to handicap certain snitchers
+/// relative to others.
 ///
-/// NOTE: This function acts eagerly.
-pub fn try_evict<T: Trait>(account: &T::AccountId, handicap: T::BlockNumber) -> RentOutcome {
-	try_evict_or_and_pay_rent::<T>(account, handicap, false).0
+/// NOTE this function performs eviction eagerly. All changes are read and written directly to
+/// storage.
+pub fn snitch_contract_should_be_evicted<T: Trait>(
+	account: &T::AccountId,
+	handicap: T::BlockNumber,
+) -> bool {
+	let contract_info = <ContractInfoOf<T>>::get(account);
+	let alive_contract_info = match contract_info {
+		None | Some(ContractInfo::Tombstone(_)) => return false,
+		Some(ContractInfo::Alive(contract)) => contract,
+	};
+	let current_block_number = <frame_system::Module<T>>::block_number();
+	let verdict = consider_case::<T>(
+		account,
+		current_block_number,
+		handicap,
+		&alive_contract_info,
+	);
+
+	// Enact the verdict only if the contract gets removed.
+	match verdict {
+		Verdict::Kill => {
```

Comment:
```suggestion
		Verdict::Kill | Verdict::Evict { .. } => {
```

Reply:
Done! (Although I didn't use your suggestion since the arm must be removed as well)